src/Database/LSMTree/Internal/RefCount.hs

```diff
@@ -0,0 +1,85 @@
+  , RefCount (..)
+  , unsafeMkRefCounterN
+  , mkRefCounterN
+  , mkRefCounter1
+  , addReference
+  , removeReference
+  , readRefCount
+  ) where
+
+import           Control.DeepSeq
+import           Control.Exception (assert)
+import           Control.Monad (when)
+import           Control.Monad.Class.MonadThrow
+import           Control.Monad.Primitive
+import           Data.Maybe
+import           Data.Primitive.PrimVar
+
+-- | A reference counter with an optional finaliser action. Once the reference
+-- count reaches @0@, the finaliser will be run.
+data RefCounter m = RefCounter {
+    countVar  :: !(PrimVar (PrimState m) Int)
+  , finaliser :: !(Maybe (m ()))
+  }
+
+-- | NOTE: Only strict in the variable and not the referenced value.
+instance NFData (RefCounter m) where
+  rnf RefCounter{countVar, finaliser} =
+      rwhnf countVar `seq` rwhnf finaliser
+
+newtype RefCount = RefCount Int
+  deriving stock (Eq, Ord, Show)
+
+{-# SPECIALISE unsafeMkRefCounterN :: RefCount -> Maybe (IO ()) -> IO (RefCounter IO) #-}
+-- | Like 'mkRefCounterN', but throws an error if the initial @n < 1@
+unsafeMkRefCounterN :: PrimMonad m => RefCount -> Maybe (m ()) -> m (RefCounter m)
+unsafeMkRefCounterN !n finaliser = mkRefCounterN n finaliser >>= \case
+    Nothing -> error "unsafeMkRefCounterN: n < 1"
+    Just rc -> pure rc
+
+{-# SPECIALISE mkRefCounterN :: RefCount -> Maybe (IO ()) -> IO (Maybe (RefCounter IO)) #-}
+-- | Make a reference counter with initial value @n@. An optional finaliser is
+-- run when the reference counter reaches @0@.
+--
+-- Returns 'Nothing' if @n < 1@, and 'Just' otherwise.
+mkRefCounterN :: PrimMonad m => RefCount -> Maybe (m ()) -> m (Maybe (RefCounter m))
+mkRefCounterN (RefCount !n) finaliser
+  | n < 1 = pure Nothing
+  | otherwise = do
+    -- evaluate the finaliser a little bit before we store it
+    let !() = case finaliser of
+                Nothing       -> ()
+                Just !_action -> ()
+    countVar <- newPrimVar $! n
+    pure $! Just $! RefCounter{countVar, finaliser}
+
+{-# SPECIALISE mkRefCounter1 :: Maybe (IO ()) -> IO (RefCounter IO) #-}
+-- | Make a reference counter with initial value @1@. An optional finaliser is
+-- run when the reference counter reaches @0@.
+mkRefCounter1 :: PrimMonad m => Maybe (m ()) -> m (RefCounter m)
+mkRefCounter1 finaliser = fromJust <$> mkRefCounterN (RefCount 1) finaliser
+
+{-# SPECIALISE addReference :: RefCounter IO -> IO () #-}
+-- | Increase the reference counter by one.
+addReference :: PrimMonad m => RefCounter m -> m ()
+addReference RefCounter{countVar} = do
+    prevCount <- fetchAddInt countVar 1
+    assert (prevCount > 0) $ pure ()
+
+{-# SPECIALISE removeReference :: RefCounter IO -> IO () #-}
+-- | Decrease the reference counter by one.
+removeReference :: (PrimMonad m, MonadMask m) => RefCounter m -> m ()
+removeReference RefCounter{countVar, finaliser} = mask_ $ do
+    prevCount <- fetchSubInt countVar 1
+    assert (prevCount > 0) $ pure ()
```

Comment:
Having read some other recent PRs, I now think I see why you were throwing an exception on adding a reference when it was at count 0, but not when removing a reference.

It's because we declare the `close` operation on tables, sessions and cursors to be idempotent. But if our close involves a ref count removal then we would need to make the ref counting "sloppy" on the removal but not the adding.

Or can we be confident that the user API of `close` doesn't cause a mismatch in referencing counting for internal objects and so we can be strict about always matching pairs of add/remove ref counts?

Reply:
`close` on tables/sessions/cursors being idempotent is mainly nice for the user. In the implementation, we only rely on idempotency to allow session `close` to occur concurrently with table/cursor `close`

I think we can be confident that the public API doesn't cause a mismatch in reference counts. There are relatively few places where references are increased/decreased, and if we handle all these increases/decreases in an exception safe way (e.g., using `TempRegistry`) then we should never leave a reference count in an off-by-one state

Reply:
If it turns out I'm wrong, then we can always revisit the idempotency question

Reply:
If necessary, it's possible instead of using atomicAdd/Sub to use a CAS loop to check the value before and then atomically inc/decrement it. But that's more complex.